scriptworker/cot/verify.py

```diff
@@ -655,23 +663,37 @@ def _sort_dependencies_by_name_then_task_id(dependencies):
         DownloadError: on failure.
 
     """
-    async_tasks = []
+    mandatory_artifact_tasks = []
+    optional_artifact_tasks = []
     # only deal with chain.links, which are previously finished tasks with
     # signed chain of trust artifacts.  ``chain.task`` is the current running
     # task, and will not have a signed chain of trust artifact yet.
     for link in chain.links:
         task_id = link.task_id
         url = get_artifact_url(chain.context, task_id, 'public/chainOfTrust.json.asc')
         parent_dir = link.cot_dir
-        async_tasks.append(
-            asyncio.ensure_future(
-                download_artifacts(
-                    chain.context, [url], parent_dir=parent_dir,
-                    valid_artifact_task_ids=[task_id]
-                )
+        coroutine = asyncio.ensure_future(
+            download_artifacts(
+                chain.context, [url], parent_dir=parent_dir,
+                valid_artifact_task_ids=[task_id]
             )
         )
-    paths = await raise_future_exceptions(async_tasks)
+
+        if is_task_required_by_any_mandatory_artifact(chain, task_id):
+            mandatory_artifact_tasks.append(coroutine)
+        else:
+            optional_artifact_tasks.append(coroutine)
+
+    mandatory_artifacts_paths = await raise_future_exceptions(mandatory_artifact_tasks)
+    succeeded_optional_artifacts_paths, failed_optional_artifacts = \
+        await get_results_and_future_exceptions(optional_artifact_tasks)
```

Comment:
We're going to download in two phases, which is probably fine. I'm not sure how to improve that without adding too much complexity.

Reply:
Right. Maybe we could define a function like `raise_future_exceptions(mandatory_artifact_tasks, optionals=optional_artifact_tasks)`. It would perform the download and sort what completed from what's not. It can be done in a follow up, if you wish.